test/test_basic_extractor.py

```diff
@@ -0,0 +1,171 @@
+        super().__init__()
+        self.num_layers = num_layers
+        self.hidden_size = hidden_size
+        self.all_tokens = all_tokens
+        self.all_activations = all_activations
+        self.all_pairs = None
+        self.reset()
+
+    @overrides
+    def forward(self, token, activations):
+        # Consume next activation, make sure it's the right token
+        next_token, next_activation = next(self.all_pairs)
+        assert token == next_token
+
+        return None, {0: {"hx": next_activation, "cx": next_activation}}
+
+    def reset(self):
+        """ Reset the activations for next test. """
+        self.all_pairs = zip(self.all_tokens, self.all_activations)
+
+
+class TestExtractor(unittest.TestCase):
+    """ Test functionalities of the Extractor class. """
+
+    @classmethod
+    def setUpClass(self):
+        # Prepare Mock sentences
+        self.test_sentences = [MagicMock(), MagicMock(), MagicMock()]
+        self.test_sentences[0].sen = ["The", "ripe", "taste", "improves", "."]
+        self.test_sentences[0].labels = [0, 0, 1, 0, 0]
+        self.test_sentences[0].misc_info = {"quality": "delicious"}
+
+        self.test_sentences[1].sen = ["The", "hog", "crawled", "."]
+        self.test_sentences[1].labels = [0, 1, 0, 0]
+        self.test_sentences[1].misc_info = {"quality": "hairy"}
+
+        self.test_sentences[2].sen = ["Move", "the", "vat", "."]
+        self.test_sentences[2].labels = [0, 0, 1, 0]
+        self.test_sentences[2].misc_info = {"quality": "ok"}
+
+        self.corpus = {i: self.test_sentences[i] for i in range(len(self.test_sentences))}
+
+        # Mock the activations the model produces
+        self.all_tokens = list(itertools.chain(*[sentence.sen for sentence in self.test_sentences]))
+
+        self.test_sentence_activations = []
+        identifier_value = 0
+        for sentence in self.corpus.values():
+            self.test_sentence_activations.append(
+                create_sentence_dummy_activations(len(sentence.sen), ACTIVATION_DIM, identifier_value)
+            )
+            identifier_value += len(sentence.sen)
+
+        self.all_activations = torch.cat(self.test_sentence_activations)
+
+        # Prepare Mock Model
+        self.model = MockLanguageModel(
+            num_layers=1, hidden_size=ACTIVATION_DIM, all_tokens=self.all_tokens, all_activations=self.all_activations
+        )
+
+        # Init extractor
+        self.extractor = Extractor(self.model, self.corpus, ACTIVATION_NAMES, output_dir=ACTIVATIONS_DIR)
+
+    def test_extract_sentence(self):
+        """ Test the _extract_sentence class, especially with focus on the selection function functionality. """
+
+        def _merge_sentence_activations(sentences_activations):
+            return np.array(list(itertools.chain(
+                *[sentence_activations[(0, "hx")] for sentence_activations in sentences_activations])
+            ))
+
+        # Test extraction of all activations
```

Comment:
maybe it makes sense to put these different ones in different functions (test_extract_by_token, test_extract_by_position... etc)?

Reply:
Done!